File: Machine Learning A-Z - udemy - superdatascience/poly_reg_own.R
# Ploynomial Regression

# import
dataset <- read.csv("Position_Salaries.csv")
dataset <- dataset[2:3]

# we don't split the set because of the less data there are only 10 records

# fitting linear regression to the dataset
lin_reg <-  lm(formula=Salary~.,
             data=dataset)

# fitting polynomial regression
# here the degree will be 2
dataset$Level2 <- dataset$Level^2
dataset$Level3 <- dataset$Level^3
dataset$Level4 <- dataset$Level^4
dataset$Level5 <- dataset$Level^5
poly_reg <- lm(formula=Salary~.,
               data=dataset)

# visualision the lin reg result
library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level, y=dataset$Salary),
              colour= 'red')+
  geom_line(aes(x=dataset$Level, y=predict(lin_reg,newdata = dataset)),
            colour= 'blue')+
  ggtitle('Linear Regression on Level and Salary')+
  xlab('Level')+
  ylab('Salary')

# visualision the ploynomial regression result
ggplot()+
  geom_point(aes(x=dataset$Level, y=dataset$Salary),
             colour= 'red')+
  geom_line(aes(x=dataset$Level, y=predict(poly_reg,newdata = dataset)),
            colour= 'blue')+
  ggtitle('Ploynomial Regression on Level and Salary')+
  xlab('Level')+
  ylab('Salary')

# predict new result with lin_reg
y_pred <- predict(lin_reg, data.frame(Level=6.5))

# predict new result with poly_reg
y_pred_poly <- predict(poly_reg, data.frame(Level=6.5,
                                            Level2=6.5^2,
                                            Level3=6.5^3,
                                            Level4=6.5^4,
                                            Level5=6.5^5))

File: Machine Learning A-Z - udemy - superdatascience/poly_reg_own.py
# -*- coding: utf-8 -*-
"""
Created on Mon Aug  6 16:58:54 2018

@author: Polla
"""

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

dataset=pd.read_csv('Position_Salaries.csv')
# with the following code the result would be a list but we would like
# to use matrix
#X=dataset.iloc[:,1].values
# so create X as a matrix
X=dataset.iloc[:,1:2].values
y=dataset.iloc[:,2].values


# SPLITTING TRAIN AND TEST SET
# we don't split the set because we have only 10 observation
# from sklearn.model_selection import train_test_split
# X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=0)

from sklearn.linear_model import LinearRegression
lin_regressor = LinearRegression()
lin_regressor.fit(X,y)


from sklearn.preprocessing import PolynomialFeatures
poly_regressor = PolynomialFeatures(degree=2)
X_poly=poly_regressor.fit_transform(X)
lin_regressor_2 = LinearRegression()
lin_regressor_2.fit(X_poly, y)


poly_regressor_3 = PolynomialFeatures(degree=3)
X_poly_3=poly_regressor_3.fit_transform(X)
lin_regressor_3 = LinearRegression()
lin_regressor_3.fit(X_poly_3, y)

# Visualising the linear regression
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_regressor.predict(X), color = 'blue')
plt.title('Linear Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the regression with degree=2
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_regressor_2.predict(X_poly), color = 'green')
plt.title('Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the regression with degree=3
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_regressor_3.predict(X_poly_3), color = 'green')
plt.title('Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the regression with degree=3 in more nice way
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_regressor_3.predict(poly_regressor_3.fit_transform(X_grid)), color = 'green')
plt.title('Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()


# Predicting new results
lin_regressor.predict(6.5)

lin_regressor_2.predict(poly_regressor.fit_transform(6.5))